### PR TITLE
[prometheus] lock main queue during update

### PR DESCRIPTION
**Pull request by vitaliy-sn**
Signed-off-by: Vitaliy Snurnitsin <vitaliy.snurnitsin@flant.com>

## Description

Lock main queue during deckhouse update in clusters with HA enabled.

## Why do we need it, and what problem does it solve?

When we are updating deckhouse and restarting a prometheus pod - we can left without any monitoring and alerts.

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: prometheus
type: feature
summary: Lock main queue during deckhouse update in clusters with HA enabled.
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/gcp'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/gcp'...

File: modules/300-prometheus/hooks/lock_main_queue_during_update.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	return cpod, nil
+}
+
+func handleLockMainQueue(input *go_hook.HookInput) error {
+	if !input.Values.Get("global.clusterIsBootstrapped").Bool() {
+		input.LogEntry.Info("Cluster is not yet bootstrapped, not locking main queue")
+		return nil
+	}
+
+	if !input.Values.Get("global.discovery.clusterControlPlaneIsHighlyAvailable").Bool() {
```
Comment (name212) [inline modules/300-prometheus/hooks/lock_main_queue_during_update.go:95]: we can disable high availability in the module config. Do we have to consider this case too?

File: modules/300-prometheus/hooks/lock_main_queue_during_update.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	cpod := prometheusPod{
+		Name:    pod.Name,
+		IsReady: isReady,
+	}
+
+	return cpod, nil
+}
+
+func handleLockMainQueue(input *go_hook.HookInput) error {
+	if !input.Values.Get("global.clusterIsBootstrapped").Bool() {
```
Comment (name212) [inline modules/300-prometheus/hooks/lock_main_queue_during_update.go:90]: Let's imagine the situation: we set up a cluster with one static node (frontend) and one system node.
The frontend node is bootstrapped before the system node. The node is created, the cluster is considered bootstrapped. Then this hook executes, sees that the cluster is bootstrapped, but there are no pods and locates the main queue. Question: will the system node be created?

File: modules/300-prometheus/hooks/lock_main_queue_during_update_test.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+			f.BindingContexts.Set(f.KubeStateSet(runningNotReadyPods))
+			f.ValuesSet("global.discovery.clusterControlPlaneIsHighlyAvailable", false)
+			f.RunHook()
+		})
+
+		It("Must be executed successfully", func() {
+			Expect(f).To(ExecuteSuccessfully())
+		})
+
+	})
```
Comment (name212) [inline modules/300-prometheus/hooks/lock_main_queue_during_update_test.go:134]: I think we need some tests:
- when cluster is not bootstrapped
- prometheus is not in HA mode

**Review by name212 — COMMENTED** [review]
tests and questions

**Review by name212 — CHANGES_REQUESTED** [review]
questions and tests

File: modules/300-prometheus/hooks/lock_main_queue_during_update.go (hunk 1 above)
Comment (vitaliy-sn) [reply modules/300-prometheus/hooks/lock_main_queue_during_update.go]: I think yes. Added high availability check from module config.

**Pull request closed without merging**